core/src/main/java/software/amazon/awssdk/retry/RetryHandler.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.retry;
+
+import static java.util.Collections.singletonList;
+
+import java.time.Duration;
+import java.util.Optional;
+import software.amazon.awssdk.RequestExecutionContext;
+import software.amazon.awssdk.SdkBaseException;
+import software.amazon.awssdk.http.HttpResponse;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.util.CapacityManager;
+
+public class RetryHandler {
+
+    public static final String HEADER_SDK_RETRY_INFO = "amz-sdk-retry";
+
+    private final RetryPolicy retryPolicy;
+    private final CapacityManager retryCapacity;
+
+    private Duration lastBackoffDelay = Duration.ZERO;
+    private boolean retryCapacityConsumed;
+    private RetryPolicyContext retryPolicyContext;
+    private Optional<SdkBaseException> lastRetriedException = Optional.empty();
+
+    public RetryHandler(RetryPolicy retryPolicy,
+                        CapacityManager retryCapacity) {
+        this.retryPolicy = retryPolicy;
+        this.retryCapacity = retryCapacity;
+    }
+
+    public boolean shouldRetry(HttpResponse httpResponse,
+                               SdkHttpFullRequest request,
+                               RequestExecutionContext context,
+                               SdkBaseException exception,
+                               int requestCount) {
+
+        final int retriesAttempted = requestCount - 1;
+
+        // Do not use retry capacity for throttling exceptions
+        if (!RetryUtils.isThrottlingException(exception)) {
+            // See if we have enough available retry capacity to be able to execute this retry attempt.
+            if (!retryCapacity.acquire(SdkDefaultRetrySettings.THROTTLED_RETRY_COST)) {
```

Comment:
This confused me at first. I was thinking the cost was specific to throttling exceptions but we don't acquire for throttling exceptions. I see now it's just the default cost for "retry throttling" but the naming really confused me. How about RETRY_THROTTLING_COST?